util/src/overlaydb.rs

```diff
@@ -40,340 +47,167 @@ use kvdb::{Database, DBTransaction};
 pub struct OverlayDB {
 	overlay: MemoryDB,
 	backing: Arc<Database>,
+	mode: DeletionMode,
 }
 
 impl OverlayDB {
-	/// Create a new instance of OverlayDB given a `backing` database.
-	pub fn new(backing: Database) -> OverlayDB { Self::new_with_arc(Arc::new(backing)) }
+	/// Create a new instance of OverlayDB given a `backing` database and deletion mode.
+	pub fn new(backing: Database, mode: DeletionMode) -> Self {
+		OverlayDB::new_with_arc(Arc::new(backing), mode)
+	}
 
-	/// Create a new instance of OverlayDB given a `backing` database.
-	pub fn new_with_arc(backing: Arc<Database>) -> OverlayDB {
-		OverlayDB{ overlay: MemoryDB::new(), backing: backing }
+	/// Create a new instance of OverlayDB given a shared `backing` database.
+	pub fn new_with_arc(backing: Arc<Database>, mode: DeletionMode) -> Self {
+		OverlayDB {
+			overlay: MemoryDB::new(),
+			backing: backing,
+			mode: mode
+		}
 	}
 
 	/// Create a new instance of OverlayDB with an anonymous temporary database.
-	pub fn new_temp() -> OverlayDB {
-		let mut dir = env::temp_dir();
+	pub fn new_temp(mode: DeletionMode) -> Self {
+		let mut dir = ::std::env::temp_dir();
 		dir.push(H32::random().hex());
-		Self::new(Database::open_default(dir.to_str().unwrap()).unwrap())
+		Self::new(Database::open_default(dir.to_str().unwrap()).unwrap(), mode)
 	}
 
-	/// Commit all operations to given batch.
+	/// Commit all operations to given batch. Returns the number of insertions
+	/// and deletions.
 	pub fn commit_to_batch(&mut self, batch: &DBTransaction) -> Result<u32, UtilError> {
 		let mut ret = 0u32;
 		let mut deletes = 0usize;
-		for i in self.overlay.drain().into_iter() {
-			let (key, (value, rc)) = i;
-			if rc != 0 {
-				match self.payload(&key) {
-					Some(x) => {
-						let (back_value, back_rc) = x;
-						let total_rc: i32 = back_rc as i32 + rc;
-						if total_rc < 0 {
-							return Err(From::from(BaseDataError::NegativelyReferencedHash(key)));
-						}
-						deletes += if self.put_payload_in_batch(batch, &key, (back_value, total_rc as u32)) {1} else {0};
-					}
-					None => {
-						if rc < 0 {
-							return Err(From::from(BaseDataError::NegativelyReferencedHash(key)));
-						}
-						self.put_payload_in_batch(batch, &key, (value, rc as u32));
+		for (key, (value, rc)) in self.overlay.drain() {
+			match rc {
+				0 => continue,
+				1 => try!(batch.put(&key, &value)),
```

Comment:
my mistake: it's only impossible if the given batch is empty.

If one of the keys is already present in the batch or backing database, then it will be overwritten. 
I think this functionality is only used by `commit`, so it wouldn't hurt to make it private and preserve the batch invariant at least. but it would have little effective change on stuff that already exists in the database.